Doc improvements

robertbastian opened the pull request:
<!--
Thank you for your pull request to ICU4X!

Reminder: try to use [Conventional Comments](https://conventionalcomments.org/) to make comments clearer.

Please see https://github.com/unicode-org/icu4x/blob/main/CONTRIBUTING.md for general
information on contributing to ICU4X.
-->

Manishearth commented on docs/tutorials/intro.md:
```diff
@@ -129,19 +125,15 @@ Users are also free to design their own providers that best fit into their ecosy
+We're going to extend our app to use the `icu::datetime` component to format a date and time.
 
-First, we need to register our choice of the provider in `./Cargo.toml`, which can be done by running:
+First, we need to register our choice of the provider, `icu_testdata`, in `Cargo.toml`, which can be done by running:
```
issue: I think we should still explicitly mention that testdata is just test data we ship with for experimentation purposes

Manishearth commented on docs/tutorials/intro.md:
```diff
@@ -153,30 +145,42 @@ icu = "1.0"
+const LOCALE: Locale = locale!("ja"); // let's try some other language
+
 fn main() {
+    let provider = icu_testdata::any();
```
praise: good call in splitting it out

Manishearth commented on docs/tutorials/intro.md:
```diff
@@ -200,19 +204,54 @@ Here's an internationalized date!
 $ cargo add icu --features serde
-$ cargo add icu_provider_fs
-$ cargo add icu_provider --features deserialize_json
+$ cargo add icu_provider_blob
```
issue: you also need `--features deserialize_json` or `--features deserialize_postcard_1` with `cargo add icu_provider`

robertbastian replied:
Not with blob provider, that activates postcard itself.

Manishearth commented on docs/tutorials/intro.md:
```diff
@@ -200,19 +204,54 @@ Here's an internationalized date!
+* `--cldr-tag` informs the program which CLDR version to use
+* `--icuexport-tag` informs the program which ICU-exported data version to use
+* `--out` is the location where we want the generated ICU4X data to be stored
+* `--format blob` to set the format of the output (we'll discuss formats later)
```
if we're doing this we should also do `--syntax postcard` imo

robertbastian replied:
It's the default and I don't want to overcomplicate it

robertbastian replied:
Actually it's not just the default, the `syntax` flag has no effect for blob

Manishearth reviewed (changes requested):
mostly looks good

wondering if we should have a separate data tutorial, and have this tutorial _only_ use testdata
fine for now though

robertbastian commented:
I split off the data generation into a separate tutorial, added data slicing, and explored all providers in more depths. I have also actually run the code for everything, and will send a PR to fix datagen 😒

Manishearth commented:
Oh, I wasn't intending for you to do that split right now, but that works too i guess.

Manishearth commented on docs/tutorials/data_management.md:
```diff
@@ -0,0 +1,243 @@
+
+The `--keys-for-bin` argument tells `icu4x-datagen` to analyze the binary and only include keys that are used by its code. In addition, we know that we only need data for the Japanese locale. This significantly reduces the blob's file size, to 54KB, and our program still works. Quite the improvement!
+
+But there is more to optimize. You might have noticed this in the output of the `icu4x-datagen` invocation, which lists 21 keys, including clearly irrelevant ones like `datetime/ethopic/datesymbols@1`. Remember how we had to convert our `DateTime<Gregorian>` into a `DateTime<AnyCalendar>` in order to use the `DateTimeFormatter`? Turns out, using `DateTimeFormatter` pulls in data for all calendar  ...
```
```suggestion
But there is more to optimize. You might have noticed this in the output of the `icu4x-datagen` invocation, which lists 21 keys, including clearly irrelevant ones like `datetime/ethopic/datesymbols@1`. Remember how we had to convert our `DateTime<Gregorian>` into a `DateTime<AnyCalendar>` in order to use the `DateTimeFormatter`? Turns out, using `DateTimeFormatter` with `--keys-for-bin` pulls in data for all calendar types that it supports. 
```

DTF itself doesn't pull in extra data

Manishearth replied:
IMO we should also mention that the keys can be manually specified. We don't have much docs for how to pick keys, but it's worth mentioning that it's possible and saying better docs on that will come later.

robertbastian replied:
I mention it off-hand in the next paragraph. I'll add a paragraph to say look at `--help`

robertbastian replied:
Actually I didn't include it on purpose. I was going to use `--keys` for these examples, but it's a major pain to figure out which keys a constructor uses.

Manishearth replied:
Yeah I think we should _mention_ it and also say that right now it's hard to figure out which key a constructor uses beyond just using keyextract, and say we plan to improve this.

I've done this kind of thing "you can do X! it's hard right now!" in the C++/JS docs too, it's good to give people something to explore instead of not telling them it's possible.

Manishearth reviewed (approved):
praise: This is great! Mind updating index.md as well to include this?

robertbastian commented on docs/tutorials/data_management.md:
````diff
@@ -0,0 +1,243 @@
+
+```console
+$ cargo add icu --features serde
+$ cargo add icu_provider --features deserialize_json
+$ cargo add icu_provider_fs
````
I think this is really weird. The features should be on `icu_provider_fs`, so that clients don't have to add `icu_provider` just to set the feature

Manishearth replied:
I agree, we can try to fix this in the next release?

sffc replied:
I disagree with feature duplication. It is a feature of `icu_provider` because that's the thing that exports AsDeserializingBufferProvider (which admittedly most users shouldn't know or care about). It's semantically incorrect to add the feature to `icu_provider_fs` because that crate is agnostic to the chosen serialization format. I used to be in the camp that we should duplicate features, but I've drifted away from that position.

robertbastian replied:
`icu_provider` is an internal crate for all intents and purposes. Clients will never have direct dependencies on it (except for when baking tells them exactly how to add it). `icu_provider_fs` is the crate that clients will use, and that crate should be configurable directly with features.

Manishearth commented on docs/tutorials/index.md:
```diff
@@ -15,6 +15,7 @@ Contributors should also check out [CONTRIBUTING.md].
 
  [discussions forum]: https://github.com/unicode-org/icu4x/discussions
  [intro.md]: ./intro.md
+ [data_management.md]: ./data_management.md
```
it should also be mentioned above but I'm just going to fix this myself

sffc commented on docs/tutorials/data_management.md:
```diff
@@ -0,0 +1,243 @@
+* `--icuexport-tag` selects the ICU-exported data version to use
+* `--out` is the location where we want the generated ICU4X data to be stored
+* `--format` sets the format of the output (we'll discuss formats later)
+* `--all-keys` `--all-locales` specifies that we want to include all data for all locales
```
Suggestion: add "This will generate a lot of data, which we will optimize later in the tutorial."

robertbastian replied:
Given that this is already merged, please make a PR with your proposed changes.

sffc commented on docs/tutorials/data_management.md:
```diff
@@ -0,0 +1,243 @@
+    let dtf = DateTimeFormatter::try_new_with_buffer_provider(&buffer_provider, &LOCALE.into(), options.into())
+        .expect("Failed to initialize DateTimeFormatter");
+
+    let date = DateTime::try_new_gregorian_datetime(2020, 10, 14, 13, 21, 28)
```
Issue: As in the other tutorial, don't mix `try_new_gregorian_datetime` and `DateTimeFormatter`

sffc commented on docs/tutorials/data_management.md:
```diff
@@ -0,0 +1,243 @@
+
+This is a common pattern in `ICU4X`, and most of our APIs are designed with data slicing in mind.
+
+Rerunning datagen awards us with a 3KB data blob, which only contains 7 data keys!
```
Nit (word choice): the object of the verb "award" is the thing being conferred or bestowed. Consider "reward" instead.

sffc commented on docs/tutorials/data_management.md:
```diff
@@ -0,0 +1,243 @@
+
+With this provider, we can use the `unstable` constructors. These are only guaranteed to work if the `BakedDataProvider` was generated with the same version of ICU4X that you are building with, but if you build the data as part of your a build pipeline, that shouldn't be a problem.
+
+You can also make the `BakedDataProvider` implement the `AnyProvider` trait, so that it can be used with `_with_any_provider` constructors. Using these constructors is slightly less performant than the `unstable` ones, but, as the name suggests, stable across (minor) releases.
```
…and it doesn't perform automatic data slicing.
